cpp/src/arrow/type.h

```diff
@@ -250,15 +251,55 @@ struct ListType : public DataType {
   static char const* name() { return "list"; }
 
   std::string ToString() const override;
+
+ protected:
+  // Constructor for classes that are implemented as List Arrays.
+  ListType(const std::shared_ptr<DataType>& value_type, Type::type logical_type)
+      : DataType(logical_type) {
+    // TODO ARROW-187 this can technically fail, make a constructor method ?
+    children_ = {std::make_shared<Field>("item", value_type)};
+  }
+};
+
+// BinaryType type is reprsents lists of 1-byte values.
+struct BinaryType : public ListType {
```

Comment:
I was going to propose a refactoring to push some of the things we are currently doing with switch statements to be part of the *Type classes.  E.g.
1.  Dump the buffers in depth first manner.
2.  Type inspection methods IsPhysicalListType, IsPhysicalPrimitiveType

We still don't strictly need the inheritance for this, but feels makes it a little cleaner/DRY.  

One reason I can see for "String > Binary > Primitive" is this possibly better aligns with parquet definitions.   Are there other ones you are thinking of?